src/bin/pg_autoctl/monitor.c

```diff
@@ -1190,3 +1199,261 @@ monitor_get_notifications(Monitor *monitor)
+	if (length >= BUFSIZE)
+	{
+		log_error("default_version \"%s\" returned by monitor is %d characters, "
+				  "the maximum supported by pg_autoctl is %d",
+				  value, length, BUFSIZE - 1);
+		context->parsedOK = false;
+		return;
+	}
+
+	value = PQgetvalue(result, 0, 1);
+ 	length = strlcpy(context->version->installedVersion, value, BUFSIZE);
+	if (length >= BUFSIZE)
+	{
+		log_error("installed_version \"%s\" returned by monitor is %d characters, "
+				  "the maximum supported by pg_autoctl is %d",
+				  value, length, BUFSIZE - 1);
+		context->parsedOK = false;
+		return;
+	}
+
+	context->parsedOK = true;
+}
+
+/*
+ * monitor_extension_update executes ALTER EXTENSION ... UPDATE TO ...
+ */
+bool
+monitor_extension_update(Monitor *monitor, char *targetVersion)
+{
+	PGSQL *pgsql = &monitor->pgsql;
+
+	return pgsql_alter_extension_update_to(pgsql,
+										   PG_AUTOCTL_MONITOR_EXTENSION_NAME,
+										   targetVersion);
+}
+
+
+/*
+ * monitor_ensure_extension_version checks that we are running a extension
+ * version on the monitor that we are compatible with in pg_autoctl. If that's
+ * not the case, we blindly try to update the extension version on the monitor
+ * to the target version we have in our default.h.
+ *
+ * NOTE: we don't check here if that's an upgrade or a downgrade, we rely on
+ * the extension's upgrade path to be free of downgrade paths (such as
+ * pgautofailover--1.2--1.1.sql).
+ */
+bool
+monitor_ensure_extension_version(Monitor *monitor)
+{
+	MonitorExtensionVersion version = { 0 };
+
+	if (!monitor_get_extension_version(monitor, &version))
+	{
+		log_fatal("Failed to check version compatibility with the monitor "
+				  "extension \"%s\", see above for details",
+				  PG_AUTOCTL_MONITOR_EXTENSION_NAME);
+		return false;
+	}
+
+	if (strcmp(version.installedVersion, PG_AUTOCTL_EXTENSION_VERSION) != 0)
+	{
+		Monitor dbOwnerMonitor = { 0 };
+
+		log_warn("This version of pg_autoctl requires the extension \"%s\" "
+				 "version \"%s\" to be installed on the monitor, current "
+				 "version is \"%s\".",
+				 PG_AUTOCTL_MONITOR_EXTENSION_NAME,
+				 PG_AUTOCTL_EXTENSION_VERSION,
+				 version.installedVersion);
+
+		/*
+		 * Ok, let's try to update the extension then.
+		 *
+		 * For that we need to connect as the owner of the database, which was
+		 * the current $USER at the time of the `pg_autoctl create monitor`
+		 * command.
+		 */
+		if (!prepare_connection_to_current_system_user(monitor,
+													   &dbOwnerMonitor))
+		{
+			log_error("Failed to upgrade extension \"%s\" to version \"%s\": "
+					  "failed prepare a connection string to the "
+					  "monitor as the database owner",
+					  PG_AUTOCTL_MONITOR_EXTENSION_NAME,
+					  PG_AUTOCTL_EXTENSION_VERSION);
+			return false;
+		}
+
+		if (!monitor_extension_update(&dbOwnerMonitor,
+									  PG_AUTOCTL_EXTENSION_VERSION))
+		{
+			log_fatal("Failed to update extension \"%s\" to version \"%s\" "
```

Comment:
we are talking about `update` and `upgrade` in the same function. Lets choose one and stick with it for the warnings in this function.

Reply:
Fixed.